File: app/src/main/java/crux/bphc/cms/models/course/Module.java
package crux.bphc.cms.models.course;

import androidx.annotation.NonNull;
import androidx.core.text.HtmlCompat;

import com.google.gson.annotations.SerializedName;

import java.util.List;

import crux.bphc.cms.R;
import crux.bphc.cms.interfaces.CourseContent;
import crux.bphc.cms.utils.FileUtils;
import io.realm.RealmList;
import io.realm.RealmObject;
import io.realm.annotations.Ignore;
import io.realm.annotations.PrimaryKey;

/**
 *
 * @author Harshit Agarwal (16-Dec-2016)
 */


public class Module extends RealmObject implements CourseContent {
    public enum Type {
        RESOURCE, FORUM, LABEL, ASSIGNMENT, FOLDER, QUIZ, URL, PAGE, DEFAULT, BOOK
    }

    @PrimaryKey
    @SerializedName("id") private int id;
    @SerializedName("instance") private int instance;
    @SerializedName("name") private String name;
    @SerializedName("url") private String url;
    @SerializedName("modicon") private String modIcon;
    @SerializedName("modname") private String modName;
    @SerializedName("description") private String description;
    @SerializedName("contents") private RealmList<Content> contents;

    private int courseSectionId;
    private boolean isUnread;

    @Ignore private Type modType;

    @SuppressWarnings("unused")
    public Module() {
        modType = Type.DEFAULT;
        contents = new RealmList<>();
    }

    public Module(Module module) {
        this.id = module.id;
        this.instance = module.instance;
        this.name = module.name;
        this.url = module.url;
        this.modIcon = module.modIcon;
        this.modName = module.modName;
        this.description = module.description;
        this.contents = new RealmList<>();

        for (Content content : module.getContents()) {
            this.contents.add(new Content(content));
        }

        this.courseSectionId = module.courseSectionId;
        this.modType = module.modType;
        this.isUnread = module.isUnread;
    }


    /**
     * @return True if module is a downloadable resource
     */
    public boolean isDownloadable() {
        return hasContents()
                && getModType() != Type.URL
                && getModType() != Type.FORUM
                && getModType() != Type.PAGE;
    }

    /**
     * @return resource id if icon available, else -1
     */
    public int getModuleIcon() {
        switch (getModType()) {
            case RESOURCE:
                if (hasContents()) {
                    Content content = getContents().first();
                    if (content == null) return -1;
                    return FileUtils.getDrawableIconFromFileName(content.getFileName());
                } else {
                    return -1;
                }
            case ASSIGNMENT:
                return R.drawable.assignment;
            case FOLDER:
                return R.drawable.folder;
            case URL:
                return R.drawable.web;
            case PAGE:
                return R.drawable.page;
            case QUIZ:
                return R.drawable.quiz;
            case FORUM:
                return R.drawable.forum;
            case BOOK:
                return R.drawable.book;
            case DEFAULT:
                return -1;
        }
        return -1;
    }

    private Type inferModuleTypefromModuleName() {
        switch (modName.toLowerCase()) {
            case "resource":
                return Type.RESOURCE;
            case "forum":
                return Type.FORUM;
            case "label":
                return Type.LABEL;
            case "assign":
                return Type.ASSIGNMENT;
            case "folder":
                return Type.FOLDER;
            case "quiz":
                return Type.QUIZ;
            case "url":
                return Type.URL;
            case "page":
                return Type.PAGE;
            default:
                return Type.DEFAULT;
        }
    }

    public boolean isUnread() {
        return isUnread;
    }

    public void setIsUnread(boolean isUnread) {
        this.isUnread = isUnread;
    }

    public String getDescription() {
        return description;
    }

    public Type getModType() {
        if (modType == Type.DEFAULT)
            modType = inferModuleTypefromModuleName();
        return modType;
    }

    public int getId() {
        return id;
    }

    public String getUrl() {
        return url;
    }

    public String getName() {
        return HtmlCompat.fromHtml(name, HtmlCompat.FROM_HTML_MODE_COMPACT).toString().trim();
    }

    public int getInstance() {
        return instance;
    }

    public String getModIcon() {
        return modIcon;
    }

    public void setCourseSectionId(int courseSectionId) {
        this.courseSectionId = courseSectionId;
    }

    @NonNull
    public RealmList<Content> getContents() {
        if (contents == null) contents = new RealmList<>();
        return contents;
    }

    public void setContents(@NonNull List<Content> contents) {
        this.contents.addAll(contents);
    }

    public boolean hasContents() {
        return getContents().size() != 0;
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof Module && ((Module) obj).getId() == id;
    }

    @Override
    public int hashCode() {
        return id;
    }
}

File: app/src/main/java/crux/bphc/cms/models/course/Module.kt
package crux.bphc.cms.models.course

import androidx.core.text.HtmlCompat
import com.google.gson.annotations.SerializedName
import crux.bphc.cms.R
import crux.bphc.cms.interfaces.CourseContent
import crux.bphc.cms.models.UserAccount
import crux.bphc.cms.utils.FileUtils
import io.realm.RealmList
import io.realm.RealmObject
import io.realm.annotations.Ignore
import io.realm.annotations.PrimaryKey
import java.util.*
import java.util.regex.Matcher
import java.util.regex.Pattern

/**
 * @author Harshit Agarwal (16-Dec-2016)
 * @author Abhijeet Viswa
 */
open class Module(
    @PrimaryKey @SerializedName("id") var id: Int = 0,
    @SerializedName("instance") var instance: Int = 0,
    name: String = "",
    @SerializedName("url") var url: String = "",
    @SerializedName("modicon") var modIcon: String = "",
    @SerializedName("modname") private var modName: String = "",
    description: String  = "",
    @SerializedName("contents") var contents: RealmList<Content> = RealmList(),
    var courseSectionId: Int = 0,
    var isUnread: Boolean = false,
) : RealmObject(), CourseContent {
    enum class Type {
        RESOURCE, FORUM, LABEL, ASSIGNMENT, FOLDER, QUIZ, URL, PAGE, DEFAULT, BOOK
    }

    @SerializedName("name") var name: String = name
        get() {
            return HtmlCompat.fromHtml(field, HtmlCompat.FROM_HTML_MODE_COMPACT).toString()
                .trim{ it <= ' ' }
        }

    @SerializedName("description") var description: String = description
        get() {
            val pattern: Pattern = Pattern.compile(URL_REGEX)
            val matcher: Matcher = pattern.matcher(field)

            val descriptionBuffer = StringBuffer(field.length)

            while (matcher.find()) {
                val foundLink = matcher.group(1)
                val replaceWith = "<a href=\"$foundLink?token=${UserAccount.token}\">"
                matcher.appendReplacement(descriptionBuffer, replaceWith)
            }
            matcher.appendTail(descriptionBuffer)

            return descriptionBuffer.toString().trim { it <= ' ' }
        }

    @Ignore var modType: Type = Type.DEFAULT
        get() {
            if (field == Type.DEFAULT) field = inferModuleTypeFromModuleName()
            return field
        }

    val isDownloadable: Boolean
        get() = contents.isNotEmpty() && modType !in arrayOf(Type.URL, Type.FORUM, Type.PAGE)

    val moduleIcon: Int
        get() {
            return when (modType) {
                Type.RESOURCE -> if (contents.isNotEmpty()) {
                    val content = contents.first() ?: return -1
                    FileUtils.getDrawableIconFromFileName(content.fileName)
                } else {
                    -1
                }
                Type.ASSIGNMENT -> R.drawable.assignment
                Type.FOLDER -> R.drawable.outline_folder_24
                Type.URL -> R.drawable.ic_link
                Type.PAGE -> R.drawable.page
                Type.QUIZ -> R.drawable.quiz
                Type.FORUM -> R.drawable.forum
                Type.BOOK -> R.drawable.book
                Type.DEFAULT, Type.LABEL -> -1
            }
        }

    fun deepCopy(): Module = Module(
        id,
        instance,
        name,
        url,
        modIcon,
        modName,
        description,
        RealmList<Content>(*contents.map { it.copy() }.toTypedArray()),
        courseSectionId,
        isUnread,
    )

    private fun inferModuleTypeFromModuleName(): Type {
        return when (modName.lowercase(Locale.ROOT)) {
            "resource" -> Type.RESOURCE
            "forum" -> Type.FORUM
            "label" -> Type.LABEL
            "assign" -> Type.ASSIGNMENT
            "folder" -> Type.FOLDER
            "quiz" -> Type.QUIZ
            "url" -> Type.URL
            "page" -> Type.PAGE
            else -> Type.DEFAULT
        }
    }

    override fun equals(other: Any?): Boolean {
        return other is Module && other.id == id
    }

    override fun hashCode(): Int {
        return id
    }

    companion object {
        private const val URL_REGEX = "<a href=\"(.*?)\">"
    }
}